grakn-graph/src/main/java/ai/grakn/graph/internal/AbstractGraknGraph.java

```diff
@@ -285,15 +288,105 @@ private EdgeImpl addEdge(Concept from, Concept to, Schema.EdgeLabel type){
         return concepts;
     }
 
-
     ConceptLog getConceptLog() {
         ConceptLog conceptLog = localConceptLog.get();
         if(conceptLog == null){
             localConceptLog.set(conceptLog = new ConceptLog(this));
+            loadOntologyCacheIntoTransactionCache(conceptLog);
         }
         return conceptLog;
     }
 
+    private Map<TypeName, TypeImpl> getCloneCache(){
+        Map<TypeName, TypeImpl> cloneCache = localCloneCache.get();
+        if(cloneCache == null){
+            localCloneCache.set(cloneCache = new HashMap<>());
+        }
+        return cloneCache;
+    }
+
+    /**
+     * Given a concept log bound to a specific thread the central ontology cache is read into this concept log.
+     * This method performs this operation whilst making a deep clone of the cached concepts to ensure transactions
+     * do not accidentally break the central ontology cache.
+     *
+     * @param conceptLog The thread bound concept log to read the snapshot into.
+     */
+    private void loadOntologyCacheIntoTransactionCache(ConceptLog conceptLog){
+        ConcurrentMap<TypeName, TypeImpl> cachedOntologySnapshot = getCachedOntology().asMap();
+
+        //Read central cache into conceptLog cloning only base concepts. Sets clones later
+        for (TypeImpl type : cachedOntologySnapshot.values()) {
+            conceptLog.cacheConcept(clone(type));
+        }
+
+        //Iterate through cached clones completing the cloning process.
+        //This part has to be done in a separate iteration otherwise we will infinitely recurse trying to clone everything
+        for (TypeImpl<?, ?> type : getCloneCache().values()) {
+            TypeImpl centralType = cachedOntologySnapshot.get(type.getName());
+
+            if(type.isRelationType()){
+                ((RelationTypeImpl) type).copyCachedConcepts((RelationTypeImpl) centralType);
+            } else if(type.isRoleType()){
+                ((RoleTypeImpl) type).copyCachedConcepts((RoleTypeImpl) centralType);
+            } else {
+                type.copyCachedConcepts(centralType);
+            }
```

Comment:
For completeness' sake, here's why we can't use the `TypeImpl` generic `T` as the parameter: "because the cache stores everything as a *Impl you still need to do some odd casting which results in methods being bypassed again."

I would suggest that we just have a method that accepts `TypeImpl`, overridden by `RelationTypeImpl` and `RoleTypeImpl`, which will handle their logic (as well as the dirty casts necessary).